src/main/christophermanahan/github/io/carnitas/EchoServer.java

```diff
@@ -0,0 +1,31 @@
+package christophermanahan.github.io.carnitas;
+
+import java.io.IOException;
+import java.net.ServerSocket;
+import java.net.Socket;
+
+public class EchoServer implements Server {
+
+  final ServerSocket serverSocket;
+
+  EchoServer(ServerSocket serverSocket) {
+    this.serverSocket = serverSocket;
+  }
+
+  public void run() {
+    try {
+      Socket socket = serverSocket.accept();
+
+      Connection socketConnection = new SocketConnection(socket);
+      Input socketInput = new SocketInput(socket);
+      Output socketOutput = new SocketOutput(socket);
+
+      while(socketConnection.isOpen()) {
+        socketInput.receive()
+          .ifPresentOrElse(socketOutput::send, socketConnection::close);
+      }
```

Comment:
What happens if `close`, `receive`, or `send` raise exceptions here?

Reply:
I believe I will get an unhandled exception type IOException. I am guessing these should be wrapped in a try catch? I am not entirely sure how to handle exceptions yet. This is a gap in my knowledge.

Reply:
Those methods are catching `IOExceptions` though, right?

Reply:
I have now attempted to handle these exceptions more expressively with custom exceptions being thrown that attempt to continue reading and writing if possible.